Regression(255659@main) Unable to log into twitch.tv

cdumez opened the pull request:
#### 541ca5a79560c6998b3e433980a19d25b0f28fd9
<pre>
Regression(255659@main) Unable to log into twitch.tv
<a href="https://bugs.webkit.org/show_bug.cgi?id=253026">https://bugs.webkit.org/show_bug.cgi?id=253026</a>
rdar://105891522

Reviewed by Wenson Hsieh and Darin Adler.

The Twitch.tv login was failing because `screen.orientation` was returning
&quot;portrait-primary&quot; on macOS. The natural/default orientation on desktop should
be &quot;landscape-primary&quot; so this was confusing Twitch.

Also update our `screen.angle` logic to take into account the natural
orientation based on:
- <a href="https://w3c.github.io/screen-orientation/#dfn-screen-orientation-values-table">https://w3c.github.io/screen-orientation/#dfn-screen-orientation-values-table</a>

I have verified that Chrome on macOS returns &quot;landscape-primary&quot; for the type
and 0 for the angle. Our behavior is now aligned.

* LayoutTests/fast/screen-orientation/natural-orientation-expected.txt: Added.
* LayoutTests/fast/screen-orientation/natural-orientation.html: Added.
* LayoutTests/platform/ios/screen-orientation/natural-orientation-expected.txt: Added.
* Source/WebCore/page/ScreenOrientation.cpp:
(WebCore::ScreenOrientation::lock):
* Source/WebCore/page/ScreenOrientationType.h:
(WebCore::naturalScreenOrientationType):
* Source/WebCore/platform/ScreenOrientationProvider.cpp:
(WebCore::ScreenOrientationProvider::currentOrientation):
* Source/WebKit/WebProcess/WebCoreSupport/WebScreenOrientationManager.cpp:
(WebKit::WebScreenOrientationManager::currentOrientation):

Canonical link: <a href="https://commits.webkit.org/260944@main">https://commits.webkit.org/260944@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/491d27d723e58b8087ae55f35c9099d783c65aca

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110039 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19139 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42720 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/87/builds/1479 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/119073 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113991 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20606 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10332 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/102297 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115785 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/15342 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/98555 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/102297 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97303 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/30214 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/102297 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/11841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/31552 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/12463 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8512 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/17830 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/51167 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [❌ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14258 "Failed to checkout and rebase branch from PR 10760") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4116 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/page/ScreenOrientationType.h:
```diff
@@ -46,6 +46,16 @@ constexpr bool isLandscape(ScreenOrientationType type)
 
+constexpr ScreenOrientationType naturalScreenOrientationType()
+{
+#if PLATFORM(IOS_FAMILY)
+    return ScreenOrientationType::PortraitPrimary;
+#else
+    // On Desktop, the natural orientation must be landscape-primary.
+    return ScreenOrientationType::LandscapePrimary;
+#endif
```
For Apple platforms there are the following: iPhone, iPad, watchOS, tvOS, macOS. To me, it is obvious that iPhone is clearly portrait-primary. Not as clear that iPad, watchOS, and tvOS all should be.

cdumez replied:
We had this discussion a while back with Tim and Wenson. iPad apparently doesn't have a clear "natural" orientation so we had decided to stick with PortraitPrimary like iPhone.

For watchOS and tvOS, we didn't explicitly discuss it so I had lumped them all together. I don't think the screen orientation API is super relevant to these platforms. We may want to refine this at some point though.

For now at least, this patch is not changing behavior for these platforms. The behavior change is for macOS at this point, where we have a clear regression.

darinadler replied:
And presumably also changes behavior for all non-Apple platforms?

darinadler replied:
If you wrote `PLATFORM(IOS)` it would be iPhone and iPad; that’s what I would do.

cdumez replied:
Other platforms don't support this yet:
```
bool defaultShouldEnableScreenOrientationAPI()
{
#if PLATFORM(MAC)
    return true;
#elif PLATFORM(IOS_FAMILY)
    static bool shouldEnableScreenOrientationAPI = linkedOnOrAfterSDKWithBehavior(SDKAlignedBehavior::ScreenOrientationAPIEnabled);
    return shouldEnableScreenOrientationAPI;
#else
    return false;
#endif
}
```

So yes but it doesn't matter yet to them.

cdumez replied:
> If you wrote `PLATFORM(IOS)` it would be iPhone and iPad; that’s what I would do.

But why take the risk to change the behavior on Watch and TV on the branch at this point?

darinadler replied:
> But why take the risk to change the behavior on Watch and TV on the branch at this point?

OK.

cdumez replied:
I don't mind changing behavior in a follow-up if people think TV and Watch should default to landscape (For TV I could see it make sense, for watch unsure). However, in this patch I'd rather focus on the macOS regression.

othermaciej replied:
Tv’s natural orientation should be landscape, if it even supports this API, but it’s not very important, not enough to change on a branch, certainly (but maybe on trunk).

For iPad, it seems the natural orientation is increasingly evolving to be landscape, but if it already was portrait, it doesn’t seem important to change it late in a release. (Maybe for a later release).

cdumez replied:
I will follow-up for TV and iPad.

cdumez replied:
Follow-up: https://github.com/WebKit/WebKit/pull/10803

cdumez committed:
Regression(255659@main) Unable to log into twitch.tv
https://bugs.webkit.org/show_bug.cgi?id=253026
rdar://105891522

Reviewed by Wenson Hsieh and Darin Adler.

The Twitch.tv login was failing because `screen.orientation` was returning
"portrait-primary" on macOS. The natural/default orientation on desktop should
be "landscape-primary" so this was confusing Twitch.

Also update our `screen.angle` logic to take into account the natural
orientation based on:
- https://w3c.github.io/screen-orientation/#dfn-screen-orientation-values-table

I have verified that Chrome on macOS returns "landscape-primary" for the type
and 0 for the angle. Our behavior is now aligned.

* LayoutTests/fast/screen-orientation/natural-orientation-expected.txt: Added.
* LayoutTests/fast/screen-orientation/natural-orientation.html: Added.
* LayoutTests/platform/ios/screen-orientation/natural-orientation-expected.txt: Added.
* Source/WebCore/page/ScreenOrientation.cpp:
(WebCore::ScreenOrientation::lock):
* Source/WebCore/page/ScreenOrientationType.h:
(WebCore::naturalScreenOrientationType):
* Source/WebCore/platform/ScreenOrientationProvider.cpp:
(WebCore::ScreenOrientationProvider::currentOrientation):
* Source/WebKit/WebProcess/WebCoreSupport/WebScreenOrientationManager.cpp:
(WebKit::WebScreenOrientationManager::currentOrientation):

Canonical link: https://commits.webkit.org/260944@main
